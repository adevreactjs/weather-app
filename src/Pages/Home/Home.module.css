.addCity {
    margin-top: 40px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin-bottom: 40px;
}