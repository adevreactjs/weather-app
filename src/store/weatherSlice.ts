import { createSlice, PayloadAction, createAsyncThunk, AnyAction } from '@reduxjs/toolkit';
import axios from 'axios'


type Wheather = {
    id: number,
    name: string,
    main: Main
}
type Main = {
    temp: number,
    feelsLike: number,
    pressure: number,
    humidity: number
}
type WeatherListState = {
    list: Wheather[],
    hourList: HourlyWheaterDay[],
    loading: boolean,
    error: string | null,
}
type HourlyWheater = {
    lat: number,
    lon: number,
    hourly: HourlyWheaterDay[]
}
type HourlyWheaterDay = {
    dt: number,
    temp: number,
}

export const fetchWeather = createAsyncThunk<Wheather, string | any, { rejectValue: string }>(
    'weather/fetchWeather',
    async function (text, { rejectWithValue }) {
        const response = await axios.get(`https://api.openweathermap.org/data/2.5/weather?q=${text}&units=metric&appid=e630dc44d82c4116f9c2df5f62768bf7`)
        const data = await response.data;
        return data;
    }
);

export const fetchHoursWeather = createAsyncThunk<HourlyWheater, number | any, { rejectValue: number }>(
    'weather/fetchHoursWeather',
    async function (text, { rejectWithValue }) {
        const response = await axios.get(`https://api.openweathermap.org/data/2.5/onecall?lat=48.4680221&lon=35.0417711&units=metric&appid=e630dc44d82c4116f9c2df5f62768bf7`)
        const data = await response.data;
        return data;
    }
);



const initialState: WeatherListState = {
    list: [],
    hourList: [],
    loading: false,
    error: null,
}

const weatherSlice = createSlice({
    name: 'weather',
    initialState,
    reducers: {

        removeWheatherCard(state, action: PayloadAction<number | any>) {
            state.list = state.list.filter((el) => el.id !== action.payload)
        },

    },
    extraReducers: (builder) => {
        builder
            .addCase(fetchWeather.pending, (state) => {
                state.loading = true;
                state.error = null;
            })
            .addCase(fetchWeather.fulfilled, (state, action) => {
                if (state.list.some((el) => el.id === action.payload.id)) {
                    const res = state.list.map((el) => (el.id == action.payload.id ? { ...action.payload } : el));
                    state.list = res
                    state.loading = false
                    console.log('update');
                } else {
                    state.list.push(action.payload)
                    localStorage.setItem(action.payload.name, action.payload.name)
                    state.loading = false
                }

            })
            .addCase(fetchWeather.pending, (state) => {
                state.error = null;
            })
            .addCase(fetchHoursWeather.fulfilled, (state, action) => {
                // state.hourList.push(action.payload)

            })
    }
});

export const { removeWheatherCard } = weatherSlice.actions;

export default weatherSlice.reducer;