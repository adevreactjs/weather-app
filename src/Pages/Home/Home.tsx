import React, { useEffect, useState } from 'react'
import Cart from '../../Components/Cart/Cart'
import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import { useAppDispatch, useAppSelector } from '../../store/hook';
import { fetchWeather } from '../../store/weatherSlice';
import Container from '@mui/material/Container';
import CircularProgress from '@mui/material/CircularProgress';
import Grid from '@mui/material/Grid';
import cl from './Home.module.css';


const Home = () => {
  const [city, setCity] = useState<string>('')
  const { loading,error, list } = useAppSelector(state => state.weather);;
  const dispatch = useAppDispatch();

  function setCityData(e: React.ChangeEvent<HTMLInputElement>) {
    setCity(e.target.value)
  }

  function addCityCart(e: React.MouseEvent<HTMLButtonElement>) {
    dispatch(fetchWeather(city))
  }

  function getSaveLocalStorageCity() {
    Object.keys(localStorage).forEach(function (key) {
      if(key !== 'current') {
        console.log(key)
        let city = localStorage.getItem(key);
        dispatch(fetchWeather(city))
      }        
    });
  }

  useEffect(() => {
    getSaveLocalStorageCity()
  }, [])
  
  return (
    <Container maxWidth="xl">
      <div className={cl.addCity}>
        <h1>Weather City </h1>
        {error && <h2>An error occured: {error}</h2>}
        <TextField sx={{ width: '250px', marginBottom: '20px' }} id="outlined-basic" label="Город" variant="outlined" value={city} onChange={setCityData} />
        <Button sx={{ width: '250px', color: 'black', borderColor: 'black' }} variant="outlined" placeholder='City' onClick={addCityCart}>Add City</Button>
      </div>
      <Grid container spacing={{ xs: 2, md: 3 }} columns={{ xs: 4, sm: 8, md: 12 }}>
        {loading ? <Box className={cl.loader} sx={{ height: '50vh', margin: '0 auto' }}>
          <CircularProgress />
        </Box> : list.map((el) => (<Grid item xs={2} sm={4} md={4} key={el.id}>
          <Cart weather={el} key={el.id} />
        </Grid>))}
      </Grid>
    </Container>
  )
}

export default Home