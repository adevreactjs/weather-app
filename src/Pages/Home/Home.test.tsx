import { render, screen } from '@testing-library/react';
import Home from './Home';
import { Provider } from 'react-redux';
import TextField from '@mui/material/TextField';
import store from '../../store';
import userEvent from '@testing-library/user-event';

const onChange = jest.fn();
describe('Home components', () => {
  it('Home render', () => {
    const view = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(view).toMatchSnapshot();
    expect(screen.queryByRole('listitem')).toBeNull();
  });
  it('render title page', () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(screen.getByText('Weather City')).toBeInTheDocument();
  });
  it('render input', () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });
  it('render button', () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  it('work onChange', () => {
    render(<TextField value='' onChange={onChange}></TextField>);
    userEvent.type(screen.getByRole('textbox'), 'Днепр');
    expect(onChange).toHaveBeenCalledTimes(5);
  });
});
