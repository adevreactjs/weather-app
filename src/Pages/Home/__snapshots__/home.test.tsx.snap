// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home components Home render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="addCity"
        >
          <h1>
            Weather City 
          </h1>
          <div
            class="MuiFormControl-root MuiTextField-root css-c5y2py-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="outlined-basic"
              id="outlined-basic-label"
            >
              Город
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="outlined-basic"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1ftyaf0"
                >
                  <span>
                    Город
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-177h3ip-MuiButtonBase-root-MuiButton-root"
            placeholder="City"
            tabindex="0"
            type="button"
          >
            Add City
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-spacing-md-3 css-1xgjpwf-MuiGrid-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
    >
      <div
        class="addCity"
      >
        <h1>
          Weather City 
        </h1>
        <div
          class="MuiFormControl-root MuiTextField-root css-c5y2py-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Город
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Город
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-177h3ip-MuiButtonBase-root-MuiButton-root"
          placeholder="City"
          tabindex="0"
          type="button"
        >
          Add City
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-spacing-md-3 css-1xgjpwf-MuiGrid-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
